trunk/src/app/srs_app_http_stream.cpp

```diff
@@ -634,18 +634,35 @@ srs_error_t SrsLiveStream::serve_http(ISrsHttpResponseWriter* w, ISrsHttpMessage
         return srs_error_wrap(err, "http hook");
     }
 
-    alive_viewers_++;
+    // Add the viewer to the viewers list.
+    viewers_.push_back(hc);
+
+    // Serve the viewer connection.
     err = do_serve_http(w, r);
-    alive_viewers_--;
-    
+
+    // Remove viewer from the viewers list.
+    vector<ISrsExpire*>::iterator it = std::find(viewers_.begin(), viewers_.end(), hc);
+    srs_assert (it != viewers_.end());
+    viewers_.erase(it);
```

Comment:
`viewers_`, as `vector` container , has bad efficiency in `erase`,  and also `find` for unsorted data.
For one SRS instance, it is reasonable to support 1K flv http connection. 
So it's 1K unsorted vector search and delete frequently. (Not so bad, also not so good).
If the vector size is 10K, it will be serious problem.
 
 In those reason, consider `map` or `unordered_map` as the container. the `SrsHttpConn->get_id` as the key, make sure the key is unique for each `SrsHttpConn`.

Reply:
No performance issue even for 10k viewers, because this only execute once when viewer close the stream.

We should never fix problem that not exists.

Reply:
Try bellow benchmark, to randomly find a elem in 100k vector:

```cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <chrono>
using namespace std;
using namespace chrono;

int main() {
    vector<uint64_t> vec;
    for (uint64_t i = 0; i < 10 * 10000; ++i) {
        vec.push_back(i);
    }

    srand(static_cast<unsigned>(time(0)));
    for (int i = 0; i < 10; i++) {
        size_t random_index = rand() % 1000000;
        uint64_t value = vec[random_index];

        auto start = high_resolution_clock::now();
        vector<uint64_t>::iterator it = std::find(vec.begin(), vec.end(), value);
        auto end = high_resolution_clock::now();

        auto duration = duration_cast<nanoseconds>(end - start).count() / 1000.0;
        cout << "Element at index " << random_index << " is " << value << " find " << *it << endl;
        cout << "Time taken to find the element: " << duration << " microseconds" << endl << endl;
    }

    return 0;
}
```

Result:

```bash
Element at index 53239 is 53239 find 53239
Time taken to find the element: 134.25 microseconds

Element at index 147785 is 0 find 0
Time taken to find the element: 0.125 microseconds

Element at index 406623 is 0 find 0
Time taken to find the element: 0.084 microseconds

Element at index 968697 is 0 find 0
Time taken to find the element: 0.083 microseconds

Element at index 165243 is 0 find 0
Time taken to find the element: 0.041 microseconds

Element at index 884384 is 0 find 0
Time taken to find the element: 0.084 microseconds

Element at index 525581 is 0 find 0
Time taken to find the element: 0.083 microseconds

Element at index 787756 is 0 find 0
Time taken to find the element: 0.084 microseconds

Element at index 900355 is 0 find 0
Time taken to find the element: 0.042 microseconds

Element at index 102331 is 0 find 0
Time taken to find the element: 0.042 microseconds
```

Reply:
one viewer close the http connection, it exec one time, but if 10K viewers  close http flv connections, this code will run 10K times. The rarely case is the 10K viewers close the connection at same time. OK, I accept this.

Reply:
For the benchmark, The real disaster of the vector is the `erase`. I would avoid doing this as far as possible.

Reply:
Even with 10,000 viewers, each increase is only 0.01ms. When will 10,000 viewers close the connection simultaneously? We should never fix problems that do not exist. You will be completely occupied if you try to solve every imaginary problem that does not exist.

Reply:
OK, accept it.

Reply:
> For the benchmark, The real disaster of the vector is the `erase`. I would avoid doing this as far as possible.

Try this:

```cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <chrono>
using namespace std;
using namespace chrono;

int main() {
    int max_elements = 10 * 1000;

    vector<uint64_t> vec;
    for (uint64_t i = 0; i < max_elements; ++i) {
        vec.push_back(i);
    }

    srand(static_cast<unsigned>(time(0)));
    for (int i = 0; i < 10; i++) {
        size_t random_index = rand() % vec.size();
        uint64_t value = vec[random_index];

        auto start = high_resolution_clock::now();
        vector<uint64_t>::iterator it = std::find(vec.begin(), vec.end(), value);
        vec.erase(it);
        auto end = high_resolution_clock::now();

        auto duration = duration_cast<nanoseconds>(end - start).count() / 1000.0;
        cout << "Element at index " << random_index << " is " << value << " find " << *it << endl;
        cout << "Time taken to find the element: " << duration << " microseconds" << endl << endl;
    }

    return 0;
}
```

The result:

```bash
Element at index 8555 is 8555 find 8556
Time taken to find the element: 21 microseconds

Element at index 8770 is 8771 find 8772
Time taken to find the element: 25.25 microseconds

Element at index 7152 is 7152 find 7153
Time taken to find the element: 18.25 microseconds

Element at index 2981 is 2981 find 2982
Time taken to find the element: 8.5 microseconds

Element at index 2102 is 2102 find 2103
Time taken to find the element: 6.292 microseconds

Element at index 5273 is 5275 find 5276
Time taken to find the element: 13.709 microseconds

Element at index 8957 is 8963 find 8964
Time taken to find the element: 22.417 microseconds

Element at index 1154 is 1154 find 1155
Time taken to find the element: 4 microseconds

Element at index 9198 is 9206 find 9207
Time taken to find the element: 23.292 microseconds

Element at index 7305 is 7310 find 7311
Time taken to find the element: 21.625 microseconds
```

For 10,000 viewers, each increase is 0.02ms. I think the key point is to run some benchmarks and gather data when you are concerned about performance issues.